pkg/plugins/policies/meshhttproute/api/v1alpha1/meshhttproute.go

```diff
@@ -53,7 +53,44 @@ type RuleConf struct {
 	BackendRefs *[]BackendRef `json:"backendRefs,omitempty"`
 }
 
+// +kubebuilder:validation:Enum=RequestHeaderModifier;ResponseHeaderModifier
+type FilterType string
+
+const (
+	RequestHeaderModifierType  FilterType = "RequestHeaderModifier"
+	ResponseHeaderModifierType FilterType = "ResponseHeaderModifier"
+)
+
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=256
+// +kubebuilder:validation:Pattern=`^[A-Za-z0-9!#$%&'*+\-.^_\x60|~]+$`
+type HeaderName string
+
+type HeaderKeyValue struct {
+	Name  HeaderName `json:"name"`
+	Value string     `json:"value"`
+}
+
+// Only one action is supported per header name.
+// Configuration to set or add multiple values for a header must use RFC 7230
+// header value formatting, separating each value with a comma.
+type HeaderModifier struct {
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=16
+	Set []HeaderKeyValue `json:"set,omitempty"`
```

Comment:
If `Set` is optional, then:
```suggestion
	Set *[]HeaderKeyValue `json:"set,omitempty"`
```

Reply:
I don't think it needs to be a pointer since it isn't being merged, it's just one entry in the filter list, right?

Reply:
That's true, but we agreed to have all optional lists as 
```go
Field *[]Type `json:"field,omitempty"`
```
so for optional fields, you can always distinguish `nil` and `zero` input.

Reply:
But only fields optional for the purposes of merging, right? Not for every field everywhere. There must be a semantic difference between `nil` and empty list in order to do this.

For example, i could imagine merging 
consecutive header modifier filters together so that leaving nil pointer would previously set headers alone while setting an empty list would remove them, but we haven't discussed that

Reply:
We could do this and [be in line with gateway API since](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1beta1.HTTPRouteRule):

> Specifying a core filter multiple times has unspecified or implementation-specific conformance.

Reply:
No, it's not about merging, we just agree to have simple rules when choosing types:
* required field: 
    ```
    Field Type `json:"field"`
    ```
* optional field: 
    ```
    Field *Type `json:"field,omitempty"`
    ```
All mergeable fields are optional, but non-mergeable fields also can be optional.

Reply:
As far as I knew when we discussed it, it was about merging fields.

Why does this need to be a pointer? Why is a nil pointer different than an empty list?

Reply:
The reasons are not very serious but still:
* you can validate and deny the user's input `set: []`, but allow `set: null`
* it's a unified approach for all optional fields, probably we can create a simple linter based on regex

Reply:
1. but why would we do this? If we add a pointer `{set: null}` is the empty value i.e. equivalent to `{}`, isn't it? As is now, `{set: []}` is the same as `{set: null}` and `{}`. And the code is just more complicated
2. i don't see the advantage of this. why do I need a linter to check for this case when it isn't a problem?

As far as I can tell we want: `{set: []} ~ {set: null} ~ {}` for this field. Because its semantics are that it's just a set of values that should be modified.

Reply:
1. By `{set: null}` I meant `{}`. You want to allow users to not specify `set` or specify a non-empty list. But specifying `set` as an empty list `{set: []}` doesn't make sense in this case.
2. To not miss pointers when we need them. In most cases `omitempty` requires a pointer, so it'd be better to enforce this for all optional fields.

Reply:
> 1. But specifying set as an empty list {set: []} doesn't make sense in this case.

Why doesn't it make sense? All ways of specifying the empty list are equivalent and they all mean "this filter doesn't set any headers". Why increase the number of distinct inhabitants of this type for no reason? It's confusing.

Reply:
> And the code is just more complicated if we add a pointer

It's not. Just use `pointer.Deref(headerModifier.Set)` when you don't need to distinguish between `nil` and `[]`. It will return an empty slice even if `headerModifier.Set` is `nil`.

Reply:
IMO that's more complicated. We _never_ need to distinguish between `nil` and `[]`  with this field, that's why I don't want to complicate the type.

Reply:
> Why increase the number of distinct inhabitants of this type for no reason?

Because that's what you get when you make your type "Optional". You get a distinct state that means "unset". Without a pointer, there is no distinct state to show that the field is unset. 

Today in the guide the rules are simple https://github.com/kumahq/kuma/blob/master/docs/guides/new-policies.md#list. We can add another split under "Optional":
```
Do you need to distinguish between nil and zero state?
* yes
    MyListField []ItemType `json:"myListField,omitempty"`
* no
    MyListField *[]ItemType `json:"myListField,omitempty"`
```

But I don't like it and I think this makes the decision-making process more complicated (when you write policies or do a review).

Reply:
> Because that's what you get when you make your type "Optional". You get a distinct state that means "unset". Without a pointer, there is no distinct state to show that the field is unset.

I am not making a type optional. I am making a field optional. When you make a field optional, it means only that it need not be explicitly written.
Whatever value of the field's type this defaults to and what that field's type is is a separate matter. In this case, the unwritten state is semantically equivalent to the empty list.

> Without a pointer, there is no distinct state to show that the field is unset.

We do not need this for this field.

> I think this makes the decision-making process more complicated (when you write policies or do a review).

IMO better make the rarer decision making process more complicated than the using types process harder a la "code is read more than it's written"

Reply:
I think it's a generally better approach to have our Go structs closer to JSON. Because the actual API we expose is JSON. In JSON you can have both an unset list and an empty list. When you implicitly map an unset JSON list to an empty Go slice it can be not obvious to readers. 

Without a pointer, a user creates `{list: []}`, but in store, we have `{}`. I think if you send `{list: []}` you should see `{list: []}` in the store. The fact that we map JSON -> Go struct -> JSON should be only our internal implementation, it should be seen as if we don't have `Go struct` in between.

Reply:
But we already do that with any `omitempty` fields so the user can also write `{anyMergeableField: null}` and get `{}`

Reply:
I don't think we accept `null`, you need to mark a field as `// +nullable` to do so.

Reply:
It seems like we do:

```
$ echo '
apiVersion: kuma.io/v1alpha1
kind: MeshHealthCheck
metadata:
  name: zero-to-four
  namespace: kuma-system
spec:
  targetRef:
    kind: MeshService
    name: srv-000_kuma-test_svc_80
  to:
    - targetRef:
        kind: MeshService
        name: srv-004_kuma-test_svc_80
      default:
        interval: 10s
        timeout: 2s
        unhealthyThreshold: 3
        healthyThreshold: 1
        tcp: {}
        http: null' | kubectl apply -f -
meshhealthcheck.kuma.io/zero-to-four created
$ kubectl get meshhealthchecks.kuma.io -n kuma-system zero-to-four -o yaml
apiVersion: kuma.io/v1alpha1
kind: MeshHealthCheck
metadata:
  name: zero-to-four
  namespace: kuma-system
  resourceVersion: "982"
  uid: 50c43a2a-2512-42c4-b82d-cd54cefbe027
spec:
  targetRef:
    kind: MeshService
    name: srv-000_kuma-test_svc_80
  to:
  - default:
      healthyThreshold: 1
      interval: 10s
      tcp: {}
      timeout: 2s
      unhealthyThreshold: 3
    targetRef:
      kind: MeshService
      name: srv-004_kuma-test_svc_80
```

And doesn't this rule go out the window as soon as you use `omitempty` anywhere `{field: 0} => {}`?. Also as soon as we have a default because `{}` => `{fieldWithDefault: 0}`.
And it certainly doesn't seem to be [a rule that the kubernetes resources follow](https://github.com/kubernetes/apimachinery/blob/release-1.23/pkg/apis/meta/v1/types.go#L254) anyway.

Reply:
Interesting, I guess `object` in OpenAPI is nullable by default, that's why `http: null` stored as `{}`. 
> And doesn't this rule go out the window as soon as you use omitempty anywhere {field: 0} => {}?

No, because optional basic types also require a pointer (even if you don't merge). So in definition you have:
```go
FieldInt *int32 `json:"fieldInt,omitempty"`
FieldBool *book `json:"fieldBool,omitempty"` 
``` 
Thanks to this fact, if a user creates `{enable: false}` in the store they'll get `{enable: false}` (not `{}`).

Reply:
Yeah but that's only with a pointer wrapper. The point is that a user already can't expect JSON in = JSON out.

Reply:
I don't understand what you mean by "omitempty is everywhere", we're building an API where `omitempty` **always** requires a pointer wrapper. That's what we have stated in the [guide](https://github.com/kumahq/kuma/blob/master/docs/guides/new-policies.md). We always can change it by opening a PR with an updated guide and discussing alternative options with the team here.

Reply:
Kubernetes resources use `omitempty` without pointers already. There simply is no guarantee that JSON in = JSON out.

Reply:
The reason we use a pointer wrapper is because of merging. That was the only justification under discussion. That's why the guide focuses so much on mergeable fields I assume.

Reply:
There's no _reason_ to force a pointer wrapper if we don't have mergeability under consideration.

Reply:
There is nothing about "mergeable" when you need to choose between 2 options:
<img width="550" alt="image" src="https://user-images.githubusercontent.com/12110261/212939163-6803da10-b773-45f2-b02b-456d5445ed9d.png">